### PR TITLE
Replace FoX with pugixml library

### PR DESCRIPTION
**Pull request by paulromano**
This pull request replaces the [FoX](https://github.com/andreww/fox) XML library with the [pugixml](http://pugixml.org/) library. As discussed in #796, the performance of FoX on very large XML files is terrible (a 140 MB XML file taking > 20 hours to load), so we need a replacement. After looking at a number of options, I settled on pugixml because it is self-contained, extremely fast, is well-maintained, and has an easy-to-use interface.

pugixml itself is a C++ library that defines a number of classes (`xml_node`, `xml_attribute`, `xml_document`, etc.). In order to use it in OpenMC, what I've done is the following:

1. First, expose key member functions of the C++ classes in C, passing and returning pointers to structs that actually store data for each class. This was very easy because all the pugixml classes have an `internal_object()` method that returns that pointer and a constructor that accepts the struct pointer. The resulting C functions are in `src/pugixml/pugixml_c.hpp`.
2. For each C function, create a Fortran procedure interface using the C interoperability feature of the F2003 standard, namely the `ISO_C_BINDING` intrinsic module. The procedure interfaces are found in `src/pugixml/pugixml_f.F90`. As you'll see there, any arguments or return values of the C functions that are pointers are matched by `type(C_PTR)` in Fortran. This includes strings, since they are pointers to chars.
3. Create Fortran derived types -- `XMLNode`, `XMLAttribute`, `XMLDocument`, and `XMLText` -- that have type-bound procedures mimicking the member functions of the C++ classes. The type-bound procedures in turn call the C functions. These derived types also appear in `src/pugixml/pugixml_f.F90`.
4. Refactor the procedures in `src/xml_interface.F90` to use the new derived types rather than FoX types. I was able to preserve the interface for the most part. There are a few minor changes, e.g. the `get_arraysize_*` functions were replaced with `node_word_count`. A few functions are effectively replaced by type-bound procedures of the new derived types.

Most of the other changes in the source are pretty straightforward. A few points worth mentioning:
- The `pugi::xml_element` and `pugi::xml_text` C++ classes actually have `as_bool()` methods that will check for 1/0, true/false, yes/no. Thus, we don't need to handle this directly in OpenMC anymore and instead can just pass a logical when getting a value. (One caveat: it doesn't seem to allow whitespace- `<run_cmfd> true </run_cmfd>` won't work).
- I was able to get rid of the `REGION_SPEC_LEN` variable that specifies the maximum length of a cell region, `<cell region="..." />`. 
- Along with that, I had to make some changes in our HDF5 interface to handle deferred-length character variables, which are used in the new XML derived types. One change is that if you pass an empty string to write to an HDF5 dataset, it won't be written at all.

Don't be alarmed by the net 14k increase in LOC. FoX alone was 52k lines so this is actually a significant decrease. For the time being, I haven't included pugixml as a git submodule since I don't envision us needing to update it with any frequency. It would be an option though since it's hosted on [github](https://github.com/zeux/pugixml).

### CODE REVIEW

File: src/cmfd_input.F90 [hunk 1]
```diff
@@ -137,118 +138,100 @@ subroutine read_cmfd_xml()
...
+    if (check_for_node(root, "norm")) then
+      call get_node_value(root, "norm", cmfd % norm)
     end if
 
     ! Set feedback logical
-    if (check_for_node(doc, "feedback")) then
-      call get_node_value(doc, "feedback", temp_str)
-      temp_str = to_lower(temp_str)
-      if (trim(temp_str) == 'true' .or. trim(temp_str) == '1') &
-           cmfd_feedback = .true.
```
Comment (wbinventor) [inline src/cmfd_input.F90]: Very nice simplifications here and elsewhere!

**Comment by cjosey** [comment]
I just did some performance testing with opendeplete.  These results are amazing.
```
38747 lines
Old time =  11.5 seconds
New time =   0.66 seconds

77538 lines
Old time =  31.6 seconds
New time =   0.82 seconds

154811 lines
Old time = 105.4 seconds
New time =   1.22 seconds

308241 lines
New time =   1.92 seconds

1347902 lines (58.5 MB)
New time =   5.51 seconds

6733502 lines (292 MB)
New time =  25.1 seconds
```
It is very clear we switched from an O(n^2) to an O(n) algorithm with this swap.  It's that, or it is disk limited so that algorithm performance isn't even relevant.

File: docs/source/devguide/user-input.rst [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+Making User Input Changes
+=========================
+
+Users are encouraged to use OpenMC's :ref:`pythonapi` to build XML files that
+the OpenMC solver then reads during the initialization phase. Thus, to modify,
+add, or remove user input options, changes must be made both within the Python
+API and the Fortran source that reads XML files produced by the Python API. The
+following steps should be followed to make changes to user input:
+
+1. Determine the Python class you need to change. For example, if you are adding
```
Comment (wbinventor) [inline docs/source/devguide/user-input.rst]: This is great!

File: docs/source/devguide/xml-parsing.rst [hunk 3]
```diff
@@ -1,118 +0,0 @@
...
-   reading is performed in the `input_xml module`_.
-
-2. Make sure that your input can be categorized as one of the datatypes from
-   `XML Schema Part 2`_ and that parsing of the data appropriately reflects
-   this. For example, for a boolean_ value, true can be represented either by
-   "true" or by "1".
-
-3. Add code to check the variable for any possible errors.
-
-A set of `RELAX NG`_ schemata exists that enables real-time validation of input
```
Comment (wbinventor) [inline docs/source/devguide/xml-parsing.rst]: Should this block on RELAXNG be included as a final step in the new "user-input.rst" file?

File: src/hdf5_interface.F90 [hunk 4]
```diff
@@ -1836,40 +1838,41 @@ subroutine read_string(buffer, obj_id, name, indep)
...
     end if
 
-    ! Get dataspace
-    call h5dget_space_f(dset_id, space_id, hdf5_err)
-
     ! Make sure buffer is large enough
     call h5dget_type_f(dset_id, filetype, hdf5_err)
-    call h5tget_size_f(filetype, size, hdf5_err)
-    if (size > len(buffer) + 1) then
+    call h5tget_size_f(filetype, n, hdf5_err)
```
Comment (wbinventor) [inline src/hdf5_interface.F90]: It is still possible to have strings that are too large to read, even with deferred length strings?

File: src/xml_interface.F90 [hunk 5]
```diff
@@ -40,492 +33,300 @@ module xml_interface
...
-    ! Check for a sub-element
-    elem_list => getChildrenByTagName(ptr, trim(node_name))
-
-    ! Get the length of the list
-    if (getLength(elem_list) == 0) then
-      found = .false.
-      return
+    ! Check if an attribute or  exists
+    attr = node%attribute(name)
+    if (attr%associated()) then
```
Comment (wbinventor) [inline src/xml_interface.F90]: What did we decide about whitespace or no whitespace between `%`?

File: src/xml_interface.F90 [hunk 6]
```diff
@@ -40,492 +33,300 @@ module xml_interface
...
-    if (node_type == ATTR_NODE) then
-      call extractDataAttribute(ptr, node_name, result_string)
-    else
-      call extractDataContent(temp_ptr, result_string)
-    end if
-
-  end subroutine get_node_array_string
-
-!===============================================================================
 ! GET_NODE_VALUE_STRING returns a single string from attr. or node
```
Comment (wbinventor) [inline src/xml_interface.F90]: This comment is no longer directly paired up with the appropriate subroutine.

File: docs/source/devguide/xml-parsing.rst (hunk 3 above)
Comment (paulromano) [inline docs/source/devguide/xml-parsing.rst]: absolutely, I forgot to add that in

File: src/hdf5_interface.F90 (hunk 4 above)
Comment (paulromano) [inline src/hdf5_interface.F90]: Nope, there should be no limitation on string length at this point. Some of the procedures implicitly impose a limit, e.g. `get_node_value_string` returns a string through an argument that is passed in, so the string can only be as big as that argument. This is why I added `node_value_string` that returns the deferred length string itself.

File: src/xml_interface.F90 (hunk 5 above)
Comment (paulromano) [inline src/xml_interface.F90]: I was handily outvoted and our style guide [now mandates](http://openmc.readthedocs.io/en/latest/devguide/styleguide.html#whitespace-in-expressions) using spaces around `%`. Thanks for keeping me honest; I'll fix this.

**Comment by nelsonag** [comment]
The code looks good to me. I'd like to build it once before merging. After that, we can welcome in our new XML parser!

**Comment by nelsonag** [comment]
Worked great, merging in! Thanks @paulromano!

**Comment by smharper** [comment]
Just had a look through this PR.  I like the look of it.

The xml parser is dead.  Long live the xml parser!

**Pull request merged**